### PR TITLE
[GEOS-9757] Return a service exception when client provided WMS dimensions are not a match

### PR DESCRIPTION
**Pull request by aaime**
[GEOS-9757](https://osgeo-org.atlassian.net/browse/GEOS-9757) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

See ticket for details

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch-specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/services/wms/webadmin.rst [hunk 1]
```diff
@@ -46,6 +46,18 @@ The Web Map Service Interface Standard (WMS) provides a simple way to request an
 
 **Bicubic** -Looks at the sixteen nearest cells and fits a smooth curve through the points to find the output value. Bicubic interpolation may both change the input value as well as place the output value outside of the range of input values. Bicubic interpolation is recommended for smoothing continuous data, but this incurs a processing performance overhead.
 
+Dimensions settings
+-------------------
+
+The WMS specification mandates to throw an ``InvalidDimensionValue`` exception when a dimension value is not valid, and the nearest match is not enabled.
```
Comment (turingtestfail) [inline doc/en/user/source/services/wms/webadmin.rst:52]: "The WMS specification mandates the throwing of an"  or "The WMS specification mandates that we throw an"

File: src/community/ogcapi/dggs/ogcapi-dggs/src/main/java/org/geoserver/ogcapi/v1/dggs/DGGSService.java [hunk 2]
```diff
@@ -578,6 +578,7 @@ protected String getItemsPath(String collectionName) {
         return new FeaturesResponse(request.getAdaptee(), response);
     }
 
+    @Deprecated
```
Comment (turingtestfail) [inline src/community/ogcapi/dggs/ogcapi-dggs/src/main/java/org/geoserver/ogcapi/v1/dggs/DGGSService.java:581]: Javadoc or comment to explain the deprecation?

File: src/wms/src/main/java/org/geoserver/wms/CustomDimensionFilterConverter.java [hunk 3]
```diff
@@ -41,82 +38,81 @@
...
+            filterBuilder.appendFilters(attributeName, endAttributeName, convertedValues);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
         }
     }
 
-    /** Maps each configured dimension to the KVP key/value found in the request */
-    private Map<String, Pair<DimensionInfo, String>> getRawDimensionValues(
-            Map<String, String> rawKVP, MetadataMap metadataMap) {
+    public Map<String, DimensionInfo> getCustomDimensions() {
```
Comment (turingtestfail) [inline src/wms/src/main/java/org/geoserver/wms/CustomDimensionFilterConverter.java:107]: Javadoc?

File: src/wms/src/main/java/org/geoserver/wms/WMS.java [hunk 4]
```diff
@@ -1756,14 +1877,95 @@ public Filter getTimeElevationToFilter(
      * @param typeInfo Feature type info instance
      * @return builded filter
      */
+    @Deprecated
```
Comment (turingtestfail) [inline src/wms/src/main/java/org/geoserver/wms/WMS.java:1880]: Deprecation JavaDoc or comment?

File: src/wms/src/main/java/org/geoserver/wms/WMS.java [hunk 5]
```diff
@@ -1756,14 +1877,95 @@ public Filter getTimeElevationToFilter(
...
+                        .filter(e -> ((DimensionInfo) e.getValue()).isEnabled())
+                        .filter(e -> !((DimensionInfo) e.getValue()).isNearestMatchEnabled())
+                        .collect(
+                                Collectors.toMap(
+                                        e -> e.getKey(), e -> (DimensionInfo) e.getValue()));
+        ReaderDimensionsAccessor accessor =
+                new ReaderDimensionsAccessor(
+                        (GridCoverage2DReader) mapLayerInfo.getCoverageReader());
+        for (Map.Entry<String, DimensionInfo> entry : dimensions.entrySet()) {
+            String key = entry.getKey().replaceFirst("dim_", "");
```
Comment (turingtestfail) [inline src/wms/src/main/java/org/geoserver/wms/WMS.java:1935]: dim_ magic string should be defined statically somewhere

File: src/wms/src/main/java/org/geoserver/wms/featureinfo/VectorBasicLayerIdentifier.java [hunk 6]
```diff
@@ -156,7 +162,8 @@ public List<FeatureCollection> identify(FeatureInfoRequestParameters params, int
         LOGGER.log(Level.FINE, q.toString());
         // let's see if we need to reproject
         if (!wms.isFeaturesReprojectionDisabled()) {
-            // reproject the features to the request CRS, this way complex feature will also be
+            // reproject the features to the getMapRequest CRS, this way complex feature will also
+            // be
```
Comment (turingtestfail) [inline src/wms/src/main/java/org/geoserver/wms/featureinfo/VectorBasicLayerIdentifier.java:166]: Formatting issues?

**Review by turingtestfail — CHANGES_REQUESTED** [review]

File: src/community/ogcapi/dggs/ogcapi-dggs/src/main/java/org/geoserver/ogcapi/v1/dggs/DGGSService.java (hunk 2 above)
Comment (aaime) [reply src/community/ogcapi/dggs/ogcapi-dggs/src/main/java/org/geoserver/ogcapi/v1/dggs/DGGSService.java:581]: Never mind, removed deprecation, leftover from an IDE suggestion.

File: src/wms/src/main/java/org/geoserver/wms/WMS.java (hunk 5 above)
Comment (aaime) [reply src/wms/src/main/java/org/geoserver/wms/WMS.java:1935]: It is, but in a very painful way: ResourceInfo.VECTOR_CUSTOM_DIMENSION_PREFIX. Will try to collect it as a local constant instead (besides, to make things even messier, "dim_" has a different meaning in GetMap, used only to prefix custom dimensions, while in the configuration we use two different prefixes for vector and raster custom dimensions... oh my).

**Comment by aaime** [comment]
Thanks, feedback hopefully addresseed.

**Review by turingtestfail — APPROVED** [review]

**Comment by aaime** [comment]
Going to backport manually to flip the exception throwing default to false.

**Pull request merged**